<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    
    
    <link rel="shortcut icon" href="img/favicon.ico">

    
    <title>SIGEL - A context-aware genomic language model for gene spatial expression imputation, pattern detection, and function discovery</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="css/base.min.css" rel="stylesheet">
    <link href="css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="css/mystyle.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body class="homepage" >

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">

            <!-- Main title -->

            
              <a class="navbar-brand" href=".">A context-aware genomic language model for gene spatial expression imputation, pattern detection, and function discovery</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fas fa-search"></i> Search
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#0-welcome-to-sigel">0 Welcome to SIGEL</a></li>
            <li class="second-level"><a href="#01-introduction">0.1 Introduction</a></li>
                
        <li class="first-level "><a href="#1-preparation">1 Preparation</a></li>
            <li class="second-level"><a href="#11-installation">1.1 Installation</a></li>
                
            <li class="second-level"><a href="#12-getting-help">1.2 Getting Help</a></li>
                
            <li class="second-level"><a href="#13-datasets">1.3 Datasets</a></li>
                
        <li class="first-level "><a href="#2-obtain-sgrs">2 Obtain SGRs</a></li>
            <li class="second-level"><a href="#21-data-preprocessing">2.1 Data preprocessing</a></li>
                
            <li class="second-level"><a href="#22-display-gene-images">2.2 Display gene images</a></li>
                
            <li class="second-level"><a href="#23-getting-sgrs">2.3 Getting SGRs</a></li>
                
            <li class="second-level"><a href="#24-gene-co-expression-information">2.4 Gene co-expression information</a></li>
                
        <li class="first-level "><a href="#3-imputing-missing-genes-in-fish-based-st-to-enhance-transcriptomic-coverage">3 Imputing missing genes in FISH-based ST to enhance transcriptomic coverage</a></li>
            <li class="second-level"><a href="#31-load-10x-sqf-memb">3.1 Load 10x-/sqf-mEmb</a></li>
                
            <li class="second-level"><a href="#32-display-gene-images">3.2 Display gene images</a></li>
                
            <li class="second-level"><a href="#33-train-sigel">3.3 Train SIGEL</a></li>
                
            <li class="second-level"><a href="#34-train-etc-gan">3.4 Train ETC-GAN</a></li>
                
            <li class="second-level"><a href="#35-imputating-gene-expression">3.5 Imputating gene expression</a></li>
                
            <li class="second-level"><a href="#36-enhanced-spatial-clustering">3.6 Enhanced spatial clustering</a></li>
                
            <li class="second-level"><a href="#37-display-ground-truth">3.7 Display ground truth</a></li>
                
            <li class="second-level"><a href="#38-display-origin-leigen-result">3.8 Display origin leigen result</a></li>
                
            <li class="second-level"><a href="#39-display-enhanced-leigen-result">3.9 Display enhanced leigen result</a></li>
                
        <li class="first-level "><a href="#4-spatial-variability-genes-detection">4 Spatial variability genes detection</a></li>
            <li class="second-level"><a href="#41-load-dataset">4.1 Load dataset</a></li>
                
            <li class="second-level"><a href="#43-calculate-svgs-score">4.3 Calculate SVGs score</a></li>
                
            <li class="second-level"><a href="#44-display-svgs">4.4 Display SVGs</a></li>
                
            <li class="second-level"><a href="#45-display-non-svgs">4.5 DIsplay non-SVGs</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="0-welcome-to-sigel">0 Welcome to SIGEL</h1>
<p>We develop the <strong>S</strong>patially <strong>I</strong>nformed <strong>G</strong>ene <strong>E</strong>mbedding <strong>L</strong>earning (<strong>SIGEL</strong>) that can simultaneously identify spatially co-expressed genes and learn semantically meaningful gene embeddings from SRT data through a pretext task of gene clustering. SIGEL first employs an image encoder to transform the spatial expression maps of genes into gene embeddings modeled by a Student’s t mixture distribution (SMM). Subsequently, a discriminatively boosted gene clustering algorithm is applied on the posterior soft assignments of genes to the mixture components, iteratively adapting the parameters of the encoder and the SMM.</p>
<div style="text-align: center;">
    <img src="logo_name.png" style="max-width: 50%; height: auto;" width="400">
</div>

<h2 id="01-introduction">0.1 Introduction</h2>
<p>The limited exploration into spatial gene co-expression within tissues has been a significant bottleneck in fully harnessing the spatial genomic context for more insightful gene representations. To bridge this gap, we introduce <strong>SIGEL</strong>, a novel few-shot, self-supervised learning model tailored for the genomic field.</p>
<p><img alt="png" src="Figure%201.png" /></p>
<p>As shown in the figure above, <strong>SIGEL</strong> generates semantically meaningful gene Representations (SGRs) by identifying spatial gene co-expression patterns. Cofunctional and enrichment analyses of SGRs endorse their utility as genomic contexts, validated through relational semantics and exploration of gene functional ontology. Three novel SGR-based methods are proposed for enhancing FISH-based spatial transcriptomics, detecting spatially variable genes, and spatial clustering. Extensive real data results affirm the superior performance of these methods, highlighting the utility of SGRs for downstream tasks.</p>
<h1 id="1-preparation">1 Preparation</h1>
<h2 id="11-installation">1.1 Installation</h2>
<p>To use SIGEL, please download code from https://github.com/WLatSunLab/SIGEL or conduct code below:</p>
<pre><code class="language-bash">git clone https://github.com/WLatSunLab/SIGEL.git
cd SIGEL/
python3 setup.py install
</code></pre>
<h2 id="12-getting-help">1.2 Getting Help</h2>
<p>For inquiries related to SIGEL's codebase or experimental setup, please feel free to post your questions in the GitHub issues section of the <a href="https://github.com/WLatSunLab/SIGEL">SIGEL repository</a>.</p>
<h2 id="13-datasets">1.3 Datasets</h2>
<p>You can access a variety of datasets for spatial gene expression analysis:</p>
<ul>
<li><strong>The Slide-seqV2 dataset of Mouse Hippocampus (<a href="https://singlecell.broadinstitute.org/single_cell/study/SCP815/sensitive-spatial-genome-wide-expression-profiling-at-cellular-resolution#study-summary">ssq-mHippo</a>)</strong></li>
<li><strong>The 10x-Visium datasets of Human Dorsolateral Prefrontal Cortex (<a href="http://spatial.libd.org/spatialLIBD">10x-hDLPFC</a>)</strong></li>
<li><strong>The 10x-Visium dataset of Human Breast Cancer (<a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1">10x-hBC</a>)</strong></li>
<li><strong>The 10x-Visium datasets of Human Middle Temporal Gyrus (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE220442">10x-hMTG</a>)</strong></li>
<li><strong>The 10x-Visium dataset of Mouse Embryo (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178636">10x-mEmb</a>)</strong></li>
<li><strong>The SeqFISH dataset of Mouse Embryo (<a href="https://crukci.shinyapps.io/SpatialMouseAtlas/">sqf-mEmb</a>)</strong></li>
</ul>
<p>Example data required for SIGEL is available <a href="https://drive.google.com/drive/folders/1C3Gk-HVYp2dQh4id8H68M9p8IWEOIut_?usp=drive_link">here</a>. Please ensure that these data are properly organized as followes:</p>
<pre><code> . &lt;SIGEL&gt;
        ├── ...
        ├── &lt;data&gt;
        │   ├── 151676_10xvisium.h5ad
        │   ├── DLPFC_matrix_151676.dat
        │   └── &lt;mEmb&gt;
        │       ├── 10x_mEmb_matrix.dat
        │       ├── sqf_mEmb_adata.h5ad
        │       └── qf_mEmb_matrix.dat
        ├── &lt;model_pretrained&gt;
        │   │
        └── ...

</code></pre>
<h1 id="2-obtain-sgrs">2 Obtain SGRs</h1>
<p>In this task, we use 10x-hDLPFC-151676 dataset to generate SGRs.</p>
<h2 id="21-data-preprocessing">2.1 Data preprocessing</h2>
<p>We follow standard preprocessing for ST data using SCANPY, removing mitochondrial and ERCC spike-in genes initially. Genes present in fewer than 10 spots are excluded. Quality control is not applied to spatial spots to maintain data integrity. Gene expression counts are normalized by library size and log-transformed. Afterward, spatial coordinates from ST data are converted into images based on quality control outcomes.</p>
<pre><code class="language-python">from src.SIGEL.SIGEL import SIGEL
import warnings
warnings.filterwarnings(&quot;ignore&quot;)

adata= SIGEL.get_data(sample_id='151676', data_type='adata')
dataset, adata = SIGEL.data_process(adata)
gene_name = adata.var.index.values
</code></pre>
<h2 id="22-display-gene-images">2.2 Display gene images</h2>
<p>Display the visual effects of 8 genes converted into images using the <code>plt.imshow()</code> function.</p>
<pre><code class="language-python">import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, 3,figsize=(15, 10), dpi=100)

for i in range(2):
    for j in range(3):
        axs[i, j].imshow(dataset[12+i+j])
        axs[i, j].set_title(f'{gene_name[12+i+j]}')

plt.tight_layout()
plt.show()
</code></pre>
<p><img alt="png" src="show_6_gene_images.png" /></p>
<h2 id="23-getting-sgrs">2.3 Getting SGRs</h2>
<pre><code class="language-python">labels, SGRs, model = SIGEL.train(dataset, pretrain=True)
</code></pre>
<pre><code>use cuda: True
Pretrain: 100%|██████████| 30/30 [02:56&lt;00:00,  5.87s/it]
model saved to model_pretrained/SIGEL.pkl.
load pretrained mae from model_pretrained/SIGEL.pkl
Clustering: 100%|██████████| 30/30 [07:24&lt;00:00, 14.81s/it]
torch.Size([16060, 32])
</code></pre>
<h2 id="24-gene-co-expression-information">2.4 Gene co-expression information</h2>
<p>The effects of clustering as an auxiliary task will be demonstrated, and it will be evaluated whether SGRs can capture gene co-expression information.</p>
<pre><code class="language-bash">import squidpy as sq
adata.var['cluster_id']=labels
name = list(adata.var_names[adata.var['cluster_id']==1])
sq.pl.spatial_scatter(adata, color=name[:8])
</code></pre>
<p><img alt="png" src="show_gene_cluster_result.png" /></p>
<h1 id="3-imputing-missing-genes-in-fish-based-st-to-enhance-transcriptomic-coverage">3 Imputing missing genes in FISH-based ST to enhance transcriptomic coverage</h1>
<p>In this section, We introduce SIGEL-enhanced-transcriptomics-coverage (SIGEL-ETC), a novel SGR-based Generative Adversarial Network (GAN) model. It assumes gene relational semantics are consistent across ST data for the same tissue, allowing it to extrapolate spatial profiles from covered to uncovered genes using comprehensive datasets like 10x Visium. Its effectiveness is demonstrated by reproducing spatial gene profiles from a mouse embryo SeqFISH dataset (<a href="https://crukci.shinyapps.io/SpatialMouseAtlas/">sqf-mEmb</a>), guided by SGRs from a 10x Visium dataset (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178636">10x-mEmb</a>).</p>
<h2 id="31-load-10x-sqf-memb">3.1 Load 10x-/sqf-mEmb</h2>
<p>We begin by loading and visualizing SeqFISH and 10x Visium data from mice embryos that share similar health statuses and developmental stages.</p>
<pre><code class="language-bash">from src.SIGEL_ETC.src.SIGEL_ETC import SIGEL_ETC
adata = SIGEL_ETC.get_data(data='sqf', data_type='adata')
adata, key_m, dataset_m = SIGEL_ETC.data_process(adata)
#key_m, dataset_m = SIGEL_ETC.get_data(data='sqf', data_type='image')
key_v, dataset_v = SIGEL_ETC.get_data(data='10x', data_type='image')
</code></pre>
<ul>
<li><code>adata</code> refers to the sqf-mEmb dataset.</li>
<li><code>key_m</code> represents the gene names in the sqf-mEmb dataset.</li>
<li><code>dataset_m</code> denotes the imagified version of the sqf-mEmb dataset.</li>
<li><code>key_v</code> is the gene name in the 10x-mEmb dataset.</li>
<li><code>dataset_v</code> refers to the imagified 10x-mEmb dataset.</li>
</ul>
<h2 id="32-display-gene-images">3.2 Display gene images</h2>
<pre><code class="language-bash">import matplotlib.pyplot as plt
fig, axs = plt.subplots(2, 4,figsize=(15, 10), dpi=100)

for i in range(2):
    for j in range(4):
        axs[i, j].imshow(dataset_m[12+i+j])
        axs[i, j].set_title(f'{key_m[12+i+j]}')

plt.tight_layout()
plt.show()
</code></pre>
<p><img alt="png" src="show_6_gene_images_seqfish.png" /></p>
<pre><code class="language-bash">import matplotlib.pyplot as plt
fig, axs = plt.subplots(2, 4,figsize=(15, 10), dpi=100)

for i in range(2):
    for j in range(4):
        axs[i, j].imshow(dataset_v[12+i+j])
        axs[i, j].set_title(f'{key_v[12+i+j]}')

plt.tight_layout()
plt.show()
</code></pre>
<p><img alt="png" src="show_6_gene_images_visium.png" /></p>
<h2 id="33-train-sigel">3.3 Train SIGEL</h2>
<p>We train SIGEL using all genes from the 10x-mEmb dataset.</p>
<pre><code class="language-python">_, model = SIGEL_ETC.train(dataset_v, pretrain=True)
model = SIGEL_ETC.load_model()
all_gmat_v2m, all_gmat_m2v = SIGEL_ETC.data_filter(key_v, dataset_v, key_m, dataset_m)

</code></pre>
<pre><code>use cuda: True
Pretraining: 100%|██████████| 30/30 [01:11&lt;00:00,  2.37s/it]
model saved to model_pretrained/SIGEL_ETC.pkl.
load pretrained mae from model_pretrained/SIGEL_ETC.pkl
Clustering: 100%|██████████| 10/10 [02:03&lt;00:00, 12.32s/it]
</code></pre>
<ul>
<li><code>model</code>: Refers to the SIGEL model that has been trained using the imagified data from the 10x-mEmb dataset.</li>
<li><code>all_gamt_v2m</code>: A dictionary mapping gene names to sqf-mEmb image pairs, derived from intersecting genes between the sqf-mEmb and 10x-mEmb datasets.</li>
<li><code>all_gamt_m2v</code>: A dictionary mapping gene names to 10x-mEmb image pairs, also derived from intersecting genes between the sqf-mEmb and 10x-mEmb datasets.</li>
</ul>
<h2 id="34-train-etc-gan">3.4 Train ETC-GAN</h2>
<p>We use 80% of the intersecting genes between the sqf-mEmb and 10x-mEmb datasets as the training set to train SIGEL-ETC.</p>
<pre><code class="language-bash">ETC_GAN, model, SGRs = SIGEL_ETC.train_ETC(adata, all_gmat_v2m, all_gmat_m2v, model)
</code></pre>
<pre><code>100%|██████████| 100/100 [00:52&lt;00:00,  1.91it/s]
</code></pre>
<h2 id="35-imputating-gene-expression">3.5 Imputating gene expression</h2>
<p>We evaluate the SIGEL-ETC on the test set.</p>
<pre><code class="language-bash">import matplotlib.pyplot as plt
import torch
import numpy as np
gene_imp = SIGEL_ETC.sqf_gen(ETC_GAN, SGRs, adata)

def prepare_and_show_image(ax, tensor_data, title, cmap='Greys', interpolation='nearest'):
    img = torch.tensor(tensor_data)
    img = img.squeeze(0).cpu().detach().numpy()
    ax.imshow(img, cmap=cmap, interpolation=interpolation)
    ax.set_title(title)
    ax.axis('off')

# Initialize plot
fig, axs = plt.subplots(3, 3, figsize=(12, 15), dpi=300)

# Data and indices
key_m = np.array(list(all_gmat_m2v.keys()))
indices = [330, 308, 325]

for row, idx in enumerate(indices):
    # Show real 10x-mEmv
    prepare_and_show_image(axs[row, 0], all_gmat_v2m[key_m[idx]], key_m[idx])

    # Show real sqf-mEmb
    prepare_and_show_image(axs[row, 1], all_gmat_m2v[key_m[idx]], f'real_{key_m[idx]}')

    # Show generated sqf-mEmb
    img_fake = gene_imp[idx]
    img_real = torch.tensor(all_gmat_m2v[key_m[idx]]).squeeze(0).cpu().detach().numpy()
    prepare_and_show_image(axs[row, 2], img_fake, f'gen_{key_m[idx]}')

plt.show()
</code></pre>
<p><img alt="png" src="show_3_gene_imputation.png" /></p>
<h2 id="36-enhanced-spatial-clustering">3.6 Enhanced spatial clustering</h2>
<pre><code class="language-bash">import squidpy as sq
from src.SIGEL_ETC.src.utils import sequential_gen
dataset_m_660, key_m_660 = sequential_gen(model, ETC_GAN, all_gmat_v2m, all_gmat_m2v)
</code></pre>
<pre><code>Shape of the non-overlapping gene set: (16569, 35, 30)
Shape of SGRs: torch.Size([16569, 64])
The first imputation finished, get 110 new genes
The second imputation finished, get 220 new genes
The third imputation finished, get 330 new genes
</code></pre>
<h2 id="37-display-ground-truth">3.7 Display ground truth</h2>
<pre><code class="language-bash">sq.pl.spatial_scatter(
    adata, color=&quot;celltype_mapped_refined&quot;, shape=None, figsize=(5, 10), dpi=500, frameon=False, legend_loc=None, save = 'grouns_truth.png', title=None
)
</code></pre>
<p><img src="ground_truth.png" width="500"></p>
<h2 id="38-display-origin-leigen-result">3.8 Display origin leigen result</h2>
<pre><code class="language-bash">from src.SIGEL_ETC.src.utils import ari_evalution
adata = adata[:,list(all_gmat_m2v.keys())]
ari_evalution(adata)
</code></pre>
<pre><code>Iteration 0: Resolution 0.55 -&gt; 20 clusters
Iteration 1: Resolution 0.775 -&gt; 22 clusters
ARI: 0.3612215326878697
WARNING: Please specify a valid `library_id` or set it permanently in `adata.uns['spatial']`
</code></pre>
<p><img src="orign_leiden.png" width="500"></p>
<h2 id="39-display-enhanced-leigen-result">3.9 Display enhanced leigen result</h2>
<pre><code class="language-bash">X_imp = np.zeros([adata.X.shape[0], len(dataset_m_660)])
for i in range(X_imp.shape[1]):
    X_imp[:, i] = dataset_m_660[i][adata.obs['array_y'], adata.obs['array_x']]
print(X_imp.shape)
</code></pre>
<pre><code class="language-bash">import pandas as pd
import anndata as ad
n_obs = X_imp.shape[1]
var_names = key_m_660[:n_obs]
var = pd.DataFrame(index=var_names)
adata_imp = ad.AnnData(X_imp, obs=adata.obs, var=var, obsm = adata.obsm, uns = adata.uns)
</code></pre>
<pre><code class="language-bash">ari_evalution(adata_imp)
</code></pre>
<pre><code>Iteration 0: Resolution 0.55 -&gt; 41 clusters
Iteration 1: Resolution 0.325 -&gt; 31 clusters
Iteration 2: Resolution 0.21250000000000002 -&gt; 22 clusters
ARI: 0.4421413788706413
WARNING: Please specify a valid `library_id` or set it permanently in `adata.uns['spatial']`
</code></pre>
<p><img src="gen_leiden.png" width="500"></p>
<h1 id="4-spatial-variability-genes-detection">4 Spatial variability genes detection</h1>
<h2 id="41-load-dataset">4.1 Load dataset</h2>
<p>```bash
 from src.SIGEL_SVG import simu_zinb
from src.SIGEL_SVG import get_svg_score
from src.SIGEL.SIGEL import SIGEL
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scanpy as sc
import warnings
warnings.filterwarnings("ignore")</p>
<p>adata= SIGEL.get_data(sample_id='151676', data_type='adata')
dataset, adata = SIGEL.data_process(adata)
gene_name = adata.var.index.values</p>
<pre><code>
## 4.2 Train SIGEL
```bash
labels, SGRs, model = SIGEL.train(dataset=dataset, pretrain=False)
</code></pre>
<h2 id="43-calculate-svgs-score">4.3 Calculate SVGs score</h2>
<pre><code class="language-bash">svg_score = get_svg_score(SGRs, dataset, adata, model, sim_times=10)
ascending_indices = np.argsort(svg_score)
descending_indices = ascending_indices[::-1]
</code></pre>
<h2 id="44-display-svgs">4.4 Display SVGs</h2>
<pre><code class="language-bash">## gene expression whitin high svg score
plot_gene = gene_name[descending_indices[:4]]

fig, axs = plt.subplots(2, 2, figsize=(8, 8))

sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[0], show=False, ax=axs[0, 0], title=plot_gene[0], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[1], show=False, ax=axs[0, 1], title=plot_gene[1], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[2], show=False, ax=axs[1, 0], title=plot_gene[2], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[3], show=False, ax=axs[1, 1], title=plot_gene[3], vmax='p99')

plt.tight_layout()
plt.savefig(&quot;svg_high_6genes.png&quot;, dpi=100, bbox_inches=&quot;tight&quot;)
plt.show()
</code></pre>
<p><img alt="png" src="output_6_0.png" /></p>
<h2 id="45-display-non-svgs">4.5 DIsplay non-SVGs</h2>
<pre><code class="language-bash">## gene expression whitin low svg score
plot_gene = gene_name[ascending_indices[2:6]]

fig, axs = plt.subplots(2, 2, figsize=(8, 8))

sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[0], show=False, ax=axs[0, 0], title=plot_gene[0], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[1], show=False, ax=axs[0, 1], title=plot_gene[1], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[2], show=False, ax=axs[1, 0], title=plot_gene[2], vmax='p99')
sc.pl.spatial(adata, img_key=&quot;hires&quot;, color=plot_gene[3], show=False, ax=axs[1, 1], title=plot_gene[3], vmax='p99')

plt.tight_layout()
plt.savefig(&quot;svg_low_6genes.png&quot;, dpi=100, bbox_inches=&quot;tight&quot;)
plt.show()
</code></pre>
<p><img alt="png" src="output_7_0.png" /></p>
<p><strong>Everything else is prepared and ready for deployment. If you have any other needs, please directly contact Wenlin Li at the email: zipging@gmail.com.</strong></p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = "."</script>
    
    <script src="js/base.js"></script>
    <script src="search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal">
                    <span aria-hidden="true">&times;</span>
                    <span class="sr-only">Close</span>
                </button>
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>

<!--
MkDocs version : 1.6.0
Build Date UTC : 2024-07-23 06:39:22.634103+00:00
-->
