<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    
    
    
    <link rel="shortcut icon" href="img/favicon.ico">

    
    <title>SIGEL - A context-aware genomic language model for gene spatial expression imputation, pattern detection, and function discovery</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="css/base.min.css" rel="stylesheet">
    <link href="css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    
    <link href="css/mystyle.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    

     
</head>

<body class="homepage" >

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">

            <!-- Main title -->

            
              <a class="navbar-brand" href=".">A context-aware genomic language model for gene spatial expression imputation, pattern detection, and function discovery</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fas fa-search"></i> Search
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#0-welcome-to-sigel">0 Welcome to SIGEL</a></li>
            <li class="second-level"><a href="#01-introduction">0.1 Introduction</a></li>
                
        <li class="first-level "><a href="#1-preparation">1 Preparation</a></li>
            <li class="second-level"><a href="#11-installation">1.1 Installation</a></li>
                
            <li class="second-level"><a href="#12-how-to-get-help-for-sigel">1.2 How to get help for SIGEL</a></li>
                
            <li class="second-level"><a href="#13-example-datasets">1.3 Example Datasets</a></li>
                
        <li class="first-level "><a href="#2-train-model-to-obtain-sigel-generated-embeddings-sgrs-via-st-dataset">2 Train model to obtain SIGEL-Generated Embeddings (SGRs) via ST dataset</a></li>
            <li class="second-level"><a href="#21-data-preprocessing">2.1 Data preprocessing</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="0-welcome-to-sigel">0 Welcome to SIGEL</h1>
<p>We develop the <strong>S</strong>patially <strong>I</strong>nformed <strong>G</strong>ene <strong>E</strong>mbedding <strong>L</strong>earning (<strong>SIGEL</strong>) that can simultaneously identify spatially co-expressed genes and learn semantically meaningful gene embeddings from SRT data through a pretext task of gene clustering. SIGEL first employs an image encoder to transform the spatial expression maps of genes into gene embeddings modeled by a Student’s t mixture distribution (SMM). Subsequently, a discriminatively boosted gene clustering algorithm is applied on the posterior soft assignments of genes to the mixture components, iteratively adapting the parameters of the encoder and the SMM.</p>
<div style="text-align: center;">
    <img src="logo_name.png" style="max-width: 50%; height: auto;" width="400">
</div>

<h2 id="01-introduction">0.1 Introduction</h2>
<p>The limited exploration into spatial gene co-expression within tissues has been a significant bottleneck in fully harnessing the spatial genomic context for more insightful gene representations. To bridge this gap, we introduce <strong>SIGEL</strong>, a novel few-shot, self-supervised learning model tailored for the genomic field.</p>
<p><img alt="png" src="Figure%201.png" /></p>
<p>As shown in the figure above, <strong>SIGEL</strong> generates semantically meaningful gene Representations (SGRs) by identifying spatial gene co-expression patterns. Cofunctional and enrichment analyses of SGRs endorse their utility as genomic contexts, validated through relational semantics and exploration of gene functional ontology. Three novel SGR-based methods are proposed for enhancing FISH-based spatial transcriptomics, detecting spatially variable genes, and spatial clustering. Extensive real data results affirm the superior performance of these methods, highlighting the utility of SGRs for downstream tasks.</p>
<h1 id="1-preparation">1 Preparation</h1>
<h2 id="11-installation">1.1 Installation</h2>
<p>To use SIGEL, please download code from https://github.com/WLatSunLab/SIGEL or conduct code below:</p>
<pre><code class="language-bash">git clone https://github.com/WLatSunLab/SIGEL.git
</code></pre>
<h2 id="12-how-to-get-help-for-sigel">1.2 How to get help for SIGEL</h2>
<p>Questions about SIGEL's code or experiments can be asked in the GitHub issue section at https://github.com/WLatSunLab/SIGEL.</p>
<h2 id="13-example-datasets">1.3 Example Datasets</h2>
<p>You can access a variety of datasets for spatial gene expression analysis:</p>
<ul>
<li><strong>Mouse Hippocampus Dataset (<a href="https://singlecell.broadinstitute.org/single_cell/study/SCP815/sensitive-spatial-genome-wide-expression-profiling-at-cellular-resolution#study-summary">ssq-mHippo</a>)</strong>: Available for detailed study.</li>
<li><strong>Human Dorsolateral Prefrontal Cortex Datasets (<a href="http://spatial.libd.org/spatialLIBD">10x-hDLPFC</a>)</strong>: Information and datasets are accessible here.</li>
<li><strong>Human Breast Cancer Dataset (<a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1">10x-hBC</a>)</strong>: Obtainable for research purposes.</li>
<li><strong>Mouse Embryo Dataset based on 10x Visium (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178636">10x-mEmb</a>)</strong>: Available for comprehensive genetic analysis.</li>
<li><strong>Mouse Embryo Dataset based on SeqFISH (<a href="https://crukci.shinyapps.io/SpatialMouseAtlas/">sqf-mEmb</a>)</strong>: Can be accessed for advanced spatial analysis.</li>
</ul>
<p>Example data required for SIGEL is available <a href="https://drive.google.com/drive/folders/1C3Gk-HVYp2dQh4id8H68M9p8IWEOIut_?usp=drive_link">here</a>. Please ensure that these data are properly organized as followes:</p>
<pre><code> . &lt;SIGEL&gt;
        ├── ...
        ├── &lt;data&gt;
        │   ├── 151676_10xvisium.h5ad
        │   ├── DLPFC_matrix_151676.dat
        │   └── &lt;mEmb&gt;
        │       ├── 10x_mEmb_matrix.dat
        │       ├── sqf_mEmb_adata.h5ad
        │       └── qf_mEmb_matrix.dat
        ├── &lt;model_pretrained&gt;
        │   │
        └── ...

</code></pre>
<h1 id="2-train-model-to-obtain-sigel-generated-embeddings-sgrs-via-st-dataset">2 Train model to obtain SIGEL-Generated Embeddings (SGRs) via ST dataset</h1>
<p>In this task, we use 10x-visium data adata to generate embeddings SGRs.</p>
<h2 id="21-data-preprocessing">2.1 Data preprocessing</h2>
<p>The original <code>adata</code> is processed and converted into an image dataset <code>dataset</code>, with the gene names also extracted as <code>gene_name</code></p>
<pre><code class="language-python">from SIGEL.src.main.SIGEL import SIGEL
import warnings
warnings.filterwarnings(&quot;ignore&quot;)
</code></pre>
<pre><code class="language-python">## get adata and image data
adata= SIGEL.get_data(sample_id='151676', data_type='adata')
dataset, adata = SIGEL.data_process(adata)
gene_name = adata.var.index.values
</code></pre>
<pre><code class="language-python">## get SGRs
z, model = SIGEL.train(dataset, pretrain=True)
</code></pre>
<p>**Everything else is prepared and ready for deployment. If you have any other needs, please directly contact Wenlin Li at the email: zipging@gmail.com. **</p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = "."</script>
    
    <script src="js/base.js"></script>
    <script src="search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal">
                    <span aria-hidden="true">&times;</span>
                    <span class="sr-only">Close</span>
                </button>
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>

<!--
MkDocs version : 1.6.0
Build Date UTC : 2024-07-13 12:19:48.848084+00:00
-->
